Support for Python 3.11

rokm opened the pull request:
A glimpse into horrors that lay ahead...

rokm committed:
setup.cfg: declare support for python 3.11

rokm committed:
tests: disable incompatible dependencies

lxml and tinyaes both seem to be incompatible with API changes
made in python 3.11. Disable their installation on python 3.11
until they are fixed.

rokm committed:
compat: introduce is_py311

rokm committed:
depend: adjust bytecode scanner for python 3.11

Python 3.11 removed CALL_FUNCTION and CALL_METHOD opcodes, replacing
them with PRECALL + CALL. For our purposes, it should be enough to
match PRECALL only (as both opcodes have same parameter, i.e., the
argument count).

In addition, the bytecode is now peppered with CACHE instructions,
which we filter out in pre-processing phase to avoid complicating
the regex rules.

Furthermore, the low bit of argument to LOAD_GLOBAL opcode must
now be ignored.

rokm committed:
depend: move bytecode cleanup to finditer() utility function

This way, the clean-up gets applied in all relevant codepaths
(e.g., previously it was missing in the scan_code_for_ctypes()
codepath).

rokm committed:
ci: enable tests on python 3.11-dev

Reorganize the test steps a bit so that the full set of libraries
for tests is installed only for non-dev python.

rokm committed:
tests: fix test_finditer

Have the test use bytestrings instead of strings.

Also assert that the bytecode string passed to bytecode.finditer()
is in fact a bytestring.

rokm committed:
Remove our fake 'site' module

Remove our fake 'site' module and use the original one instead.

From python 3.11 changelog: "In Python 3.11, the core modules
essential for Python startup are “frozen”. This means that their
code objects (and bytecode) are statically allocated by the
interpreter". Consequently, we cannot supplant 'site' with our
fake version anymore, which makes it redundant.

It has likely been redundant for a while anyway, because the
original 'site' module should be honoring the sys.flags.no_site,
sys.flags.no_user_site, and sys.flags.isolated flags. So setting
these via corresponding interpreter initialization flags in
bootloader (which we already do) should suffice to completely
disable the site module's functionality anyway.

bwoodsend commented on PyInstaller/depend/bytecode.py:
```diff
@@ -81,30 +83,80 @@ def finditer(pattern: Pattern, string):
+            out.append(code[idx])
+            out.append(code[idx+1])
+
+        return bytes(out)
```
Would this just be too much regex :upside_down_face: ?
```suggestion
        return bytecode_regex(rb"(`CACHE`.)|(..)").sub(rb"\2", code)
```

rokm replied:
If it accounts for the fact that instruction occurs only on even bytes, then I'm all for regex...

bwoodsend replied:
It does indeed! Although I'd prefer to see the ``bytecode_regex(rb"(`CACHE`.)|(..)")`` pulled out of the function so that it isn't recalculated every time.

rokm replied:
Ah, neat! Because we always match two characters, the comparisons are indeed correctly aligned.

bwoodsend commented on PyInstaller/depend/bytecode.py:
```diff
@@ -81,30 +83,80 @@ def finditer(pattern: Pattern, string):
+        ((?:`EXTENDED_ARG`.)*
+         (?:`CALL_FUNCTION_EX`|`PRECALL`).)
+    """
+    )
```
Perhaps a better way of dealing with these Python version dependent op codes would be to alias them out in `_instruction_to_regex()` like we already do for the two already in that function? Any op-codes that don't have an equivalent can return an empty character set (`b"[]"`) to no-op them. Then we should be able to stick to one definition of `_call_function_bytecode`.

rokm replied:
Hmm, I'm not particularly fond of that remapping in `_instruction_to_regex`. I think it is going to get very confusing really fast, because for pre-3.7 (which we could remove now, I suppose), we have to map CALL_METHOD to CALL_FUNCTION, but in post-3.11, we'll need to map CALL_METHOD and CALL_FUNCTION to PRECALL.

I think a better (more intuitive and easier to read) approach is to have a single `_call_function_bytecode` regex, into which the four matching groups (global function, function/method, args, call) are injected from symbolic constants. These constants could then be defined on per-version basis, and would directly spell out which opcodes are available there. 

For example, the "call" group for pre-3.7 would be `` `CALL_FUNCTION`|`CALL_FUNCTION_EX` ``, for post-3.7 `` CALL_FUNCTION`|`CALL_METHOD`|`CALL_FUNCTION_EX` ``, and for post-3.11 `` `PRECALL`|`CALL_FUNCTION_EX` ``. 

We'd need to do the same at least for function/method part (due to LOAD_METHOD being unavailable in pre-3.7), but for consistency sake, I think all four groups should be declared in a separate variable and injected into the final regex.

bwoodsend replied:
Yeah, let's go for that (including removing the remapping already in `_instruction_to_regex()`).

bwoodsend commented on .github/workflows/ci.yml:
```diff
@@ -89,6 +87,17 @@ jobs:
+        if: ${{ !endsWith(matrix.python-version, '-dev') }}
+        run: |
+          # Install dependencies for tests - libraries
+          pip install --progress-bar=off -U -r tests/requirements-libraries.txt
```
Surely these don't need both a name and a comment?

rokm replied:
Indeed they don't.

rokm committed:
depend: simplify _cleanup_bytecode_string

Use a regex to filter CACHE instructions from the bytecode.

rokm committed:
depend: rework handling of opcode differences between python versions

bwoodsend commented:
I guess the plan is just to give this a rebase every now an then until Python 3.11 is released?

rokm commented:
Yeah, or at least until python 3.11 reaches release candidate stage.

rokm committed:
depend: bytecode: add support for EXTENDED_ARG_QUICK opcode

Python 3.11.0b1 brings EXTENDED_ARG_QUICK specialization of
EXTENDED_ARG, so we need to handle both.

The revised and extended version of this commit also ensures
that all _OPCODES_* variables are used within corresponding
non-capturing groups, even the ones that contain only a single
opcode. In case we ever need to extend the content with multiple
opcodes (as was now the case with _OPCODES_EXTENDED_ARG), the
lack of enclosing group will result in problems with matching.

Also visually reorganize the regexes a bit to make groupings
easier to read; this makes it easier to verify that each opcode
variables use is indeed enclosed with its own non-capturing
group, plus it makes it easier to notice the common patterns.

cclauss commented on .github/workflows/ci.yml:
```diff
@@ -100,6 +91,23 @@ jobs:
+        id: install_libraries
+        if: ${{ !endsWith(matrix.python-version, '-dev') }}
+        run: |
+          pip install --progress-bar=off -U -r tests/requirements-libraries.txt
```
```suggestion
          pip install --progress-bar=off --upgrade -r tests/requirements-libraries.txt
```
`pip install` has options: `--user`, `--upgrade`, `--upgrade-strategy`, and `--use-pep517` so let's not force the reader to guess.

rokm replied:
Fair enough.

cclauss commented on PyInstaller/depend/bytecode.py:
```diff
@@ -21,18 +21,22 @@
+# opcode name -> opcode map
+# Python 3.11 introduced specialized opcodes that are not covered by opcode.opmap (and equivalent dis.opmap), but dis
+# has a private map of all opcodes called _all_opmap. So use the latter, if available.
+if hasattr(dis, '_all_opmap'):
```
Simplify with builtin [`getattr`](https://docs.python.org/3/library/functions.html#getattr)?

rokm replied:
Good idea!

rokm committed:
ci: use long version of command-line switches with pip install

danyeaw commented:
Python 3.11.0rc2 is now out, any help needed to finalize this?

cclauss commented:
https://www.python.org/download/pre-releases/ --> rc1
https://github.com/actions/python-versions/releases --> rc1
https://www.python.org/downloads/release/python-3110rc1/ --> Exists
https://www.python.org/downloads/release/python-3110rc2/ --> Does not exist

cclauss commented on .github/workflows/lint.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
 
     - name: Install dependencies
-      run: pip install -U flake8 wheel setuptools yapf==0.32.0 toml
+      run: pip install --upgrade flake8 wheel setuptools yapf==0.32.0 toml
```
Given https://docs.python.org/3.11/library/tomllib.html why not use https://pypi.org/project/tomli instead of https://pypi.org/project/toml on Python < 3.11 and tomllib on Python >= 3.11?
* https://peps.python.org/pep-0680

rokm replied:
The `toml` dependency is due to `yapf`, and I don't think that supports `tomllib` (yet?) nor `tomli`. 

With next `yapf` release, we'll be able to install [`yapf[pyproject]`](https://github.com/google/yapf/blob/212c5b5ad8e172d2d914ae454c121c89cccbcb35/setup.py#L84) instead...

danyeaw commented:
@cclauss Whoops you are right, I thought I saw the release announcement. It is blocked by these two issues, so hopefully it will be out soon:

https://github.com/python/cpython/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Arelease-blocker+label%3A3.11

cclauss commented:
rc2 is now available.

Safihre commented:
@rokm would you like us to test this already on our packages or only after it's merged?

rokm commented:
Feel free to test in advance of the merge. It's always good to get early feedback.

rokm commented:
Looks like there's no surprises in rc2 (or they are not covered by the basic test suite...), so I guess this is ready for review.

Safihre commented:
@rokm So far the release seems to work for us, I managed to build our application on Windows on GitHub Actions and it runs well in the basic test.
Only thing I did notice is that the `SABnzbd.pkg` file increased from 4MB to 6MB, quite a dramatic change. Is there a way for me to investigate why?

rokm commented:
> Only thing I did notice is that the `SABnzbd.pkg` file increased from 4MB to 6MB, quite a dramatic change. Is there a way for me to investigate why?

You can check the contents of `SABnzbd.pkg` with `pyi-archive_viewer`; could be related to changes made in PyInstaller 5.4, though, and not this PR...

rokm commented:
@Safihre Hmm, I thought this was going to be caused by something like `setuptools._distutills` being collected now, but that does not seem to be the case...

It seems that the size of byte-compiled modules (i.e., the .pyc files that PyInstaller collects) has increased rather significantly in 3.11... (which doesn't even surprise me, considering we had to deal with those extra CACHE instructions in our bytecode scanner).

bwoodsend commented on .github/workflows/ci.yml:
```diff
@@ -100,6 +91,23 @@ jobs:
+        if: ${{ !startsWith(matrix.os, 'windows') }}
+        run: |
+          # Run Qt-based tests with offscreen backend. This seems to break QtWebEngine tests on Windows, so
+          # apply it only here, in non-Windows path.
```
For what it's worth, macOS and Windows GA runners have displays (or something close enough to count as a display) attached so we could make this path Linux only.

bwoodsend replied:
Probably not worth changing unless it starts to cause issues later.

bwoodsend reviewed (approved):
I guess that rc2 is close enough to the real deal that I don't feel too apprehensive about bd0bf8f954abbd9ee3173a92f5106eda96a38fdb being in the develop branch.

rokm commented:
The extra size may be due to [pep-0657](https://peps.python.org/pep-0657) - we'll have to check if if the opt-out mechanism works with PyInstaller...

rokm commented:
> The extra size may be due to [pep-0657](https://peps.python.org/pep-0657) - we'll have to check if if the opt-out mechanism works with PyInstaller...

```python
# python_size_test.py
import sys
import marshal
import importlib.util

modname = sys.argv[1]
module_co = importlib.util.find_spec(modname).loader.get_code(modname)
print(f"Marshalled CO len for {module_co}: {len(marshal.dumps(module_co))}")
```

```
[rok@alpha ~]$ python3.10 python_size_test.py zipfile
Marshalled CO len for <code object <module> at 0x7f4404425bb0, file "/usr/lib64/python3.10/zipfile.py", line 1>: 60563
[rok@alpha ~]$ python3.11 python_size_test.py zipfile
Marshalled CO len for <code object <module> at 0x55e0cd68e450, file "/usr/lib64/python3.11/zipfile.py", line 1>: 117599
[rok@alpha ~]$ PYTHONNODEBUGRANGES=1 python3.11 python_size_test.py zipfile
Marshalled CO len for <code object <module> at 0x55c3de314bb0, file "/usr/lib64/python3.11/zipfile.py", line 1>: 107697
```

Oh well...

bwoodsend commented:
It appears to work out the box for me (although it looks like another micro-optimisation). 
```python
# test.py
recursive = {}
recursive["some-long-string"] = recursive
print(len(recursive["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]["some-long-string"]))
```
```fish
> for i in (seq 1000) ; cat test.py >> long.py ; end

> pyinstaller --clean --noconfirm long.py
> du --apparent-size -h dist/long/long
2.7M	dist/long/long

> PYTHONNODEBUGRANGES=1 pyinstaller --clean --noconfirm long.py
> du --apparent-size -h dist/long/long
2.1M	dist/long/long
```

Safihre commented:
> @Safihre Hmm, I thought this was going to be caused by something like `setuptools._distutills` being collected now, but that does not seem to be the case...
> 
> It seems that the size of byte-compiled modules (i.e., the .pyc files that PyInstaller collects) has increased rather significantly in 3.11... (which doesn't even surprise me, considering we had to deal with those extra CACHE instructions in our bytecode scanner).

Thanks for checking, then this is just the way it is!

danyeaw commented:
Nice one @rokm!

Safihre commented:
@rokm I actually checked, and if I set `PYTHONNODEBUGRANGES` in our release-build GitHub Action, the size is reduced by 3MB and only 1.5MB larger than the Python 3.10 version.
Our package is still fat with it's 20MB (includes also lots of non-Python stuff), but 15% improvement is still great.
